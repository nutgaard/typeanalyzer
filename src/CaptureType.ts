export enum CaptureType {
    UNKNOWN = 'UNKNOWN',
    NULL = 'NULL',
    BOOLEAN = 'BOOLEAN',
    INT = 'INT',
    DOUBLE = 'DOUBLE',
    TEXT = 'TEXT',
    LIST = 'LIST',
    OBJECT = 'OBJECT'
}