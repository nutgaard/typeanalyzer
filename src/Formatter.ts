import {CaptureType} from "./CaptureType";
import {StringBuilder} from "./utils/StringBuilder";
import {PRNG, randomInt, seededRandom} from "./Random";
import {Capture, ListCapture, ObjectCapture, PrimitiveCapture} from "./Capture";

export interface Format {
    appendType(sb: StringBuilder, type: Type): void

    fieldType(type: CaptureType, typeReference?: string): string
}

class Type {
    readonly name: string;
    readonly fields: Field[];

    constructor(name: string, fields: Field[]) {
        this.name = name;
        this.fields = fields;
    }
}

class Field {
    readonly name: string;
    readonly type: string;
    readonly nullability: boolean;

    constructor(name: string, type: string, nullability: boolean) {
        this.name = name;
        this.type = type;
        this.nullability = nullability;
    }
}

export class TypescriptFormat implements Format {
    appendType(sb: StringBuilder, type: Type): void {
        sb.appendLine(`interface ${type.name} {`)
        for (const field of type.fields) {
            sb.appendLine(`  ${field.name}: ${field.type}${field.nullability ? ' | null' : ''};`)
        }
        sb.appendLine(`}`)
    }

    fieldType(type: CaptureType, typeReference?: string): string {
        switch (type) {
            case CaptureType.UNKNOWN:
                return "any";
            case CaptureType.NULL:
                return "null";
            case CaptureType.BOOLEAN:
                return "boolean";
            case CaptureType.INT:
                return "number";
            case CaptureType.DOUBLE:
                return "number";
            case CaptureType.TEXT:
                return "string";
            case CaptureType.LIST:
                return `Array<${typeReference}>`;
            case CaptureType.OBJECT:
                return typeReference!;
        }
    }
}

export class KotlinFormat implements Format {
    appendType(sb: StringBuilder, type: Type): void {
        sb.appendLine(`data class ${type.name}(`)
        for (const field of type.fields) {
            sb.appendLine(`    val ${field.name}: ${field.type}${field.nullability ? '?' : ''},`)
        }
        sb.appendLine(`)`)
    }

    fieldType(type: CaptureType, typeReference?: string): string {
        switch (type) {
            case CaptureType.UNKNOWN:
                return "Any";
            case CaptureType.NULL:
                return "Unit";
            case CaptureType.BOOLEAN:
                return "Boolean";
            case CaptureType.INT:
                return "Int";
            case CaptureType.DOUBLE:
                return "Double";
            case CaptureType.TEXT:
                return "String";
            case CaptureType.LIST:
                return `List<${typeReference}>`;
            case CaptureType.OBJECT:
                return typeReference!;
        }
    }
}

export class Formatter {
    private format: Format;
    private rnd: PRNG;
    private nameMap = new Map<ObjectCapture, string>();

    constructor(format: Format, rnd: PRNG = seededRandom(1)) {
        this.format = format;
        this.rnd = rnd;
    }

    print(capture: Capture): string {
        const types = this.findTypes(capture);
        const sb = new StringBuilder();
        for (const type of types.values()) {
            this.format.appendType(sb, type);
            sb.appendLine("");
        }
        return sb.toString();
    }

    private findTypes(capture: Capture, current: Map<ObjectCapture, Type> = new Map()): Map<ObjectCapture, Type> {
        if (capture instanceof ObjectCapture) {
            const fieldTypes = Array.from(capture.fields.values())
                .map((it) => this.findTypes(it))
                .reduce((acc, other) => new Map([...acc, ...other]));

            current.set(capture, this.createType(capture));
            return new Map([...current, ...fieldTypes]);
        } else if (capture instanceof ListCapture) {
            return this.findTypes(capture.subtype, current);
        } else {
            return current;
        }
    }

    private createType(capture: ObjectCapture): Type {
        if (!this.nameMap.has(capture)) {
            this.nameMap.set(capture, `Generated_${randomInt(this.rnd).toString(16)}`)
        }
        const name = this.nameMap.get(capture)!;
        const fields = Array.from(capture.fields.entries())
            .map(([key, value]) => {
                let nullability: boolean;
                if (value instanceof PrimitiveCapture) nullability = value.nullable;
                else if (value instanceof ListCapture) nullability = value.nullable;
                else if (value instanceof ObjectCapture) nullability = value.nullable;
                else nullability = true;

                return new Field(key, this.createTypeName(value), nullability)
            })

        return new Type(name, fields)
    }

    private createTypeName(capture: Capture): string {
        if (capture instanceof ListCapture) return this.format.fieldType(capture.type, this.createTypeName(capture.subtype));
        else if (capture instanceof ObjectCapture) {
            if (!this.nameMap.has(capture)) {
                this.nameMap.set(capture, `Generated_${randomInt(this.rnd).toString(16)}`)
            }
            return this.nameMap.get(capture)!;
        } else {
            return this.format.fieldType(capture.type);
        }
    }
}