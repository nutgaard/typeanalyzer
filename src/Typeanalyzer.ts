import {Capture, ListCapture, NullCapture, ObjectCapture, PrimitiveCapture, UnknownCapture} from "./Capture";
import {CaptureStats} from "./CaptureStats";
import {CaptureType} from "./CaptureType";
import {Format, Formatter, FormatterConfig} from "./Formatter";

export interface Config extends FormatterConfig {
    captureValuesFor: CaptureType[];
    uniontypeThreshold: number;
}
export type CaptureLUT = { [key in CaptureType]?: boolean };
export interface CaptureConfig extends Config {
    lut: CaptureLUT;
}
export class Typeanalyzer {
    #previousCapture: Capture | null = null;
    #stats: CaptureStats = new CaptureStats();
    #config: CaptureConfig;

    constructor(config: Config = { uniontypeThreshold: 0, captureValuesFor: [] }) {
        const captureLUT: CaptureLUT = {};

        for (const type of config.captureValuesFor) {
            captureLUT[type] = true;
        }
        this.#config = { ...config, lut: captureLUT };
    }

    public get stats(): CaptureStats {
        return this.#stats
    }

    capture(value: any | undefined | null): Capture {
        try {
            const json = JSON.stringify(value);
            const obj = JSON.parse(json);
            const capture: Capture = this.createCapture(obj);
            const reconciledCapture: Capture = this.#previousCapture ? this.#previousCapture.reconcile(this.#config, capture) : capture;
            this.#stats.capture(reconciledCapture != this.#previousCapture)
            this.#previousCapture = reconciledCapture;
        } catch (e: unknown) {
            if (e instanceof Error) {
                this.#stats.error(e)
            } else {
                this.#stats.error(new Error(JSON.stringify(e)));
            }
        }
        return this.report();
    }

    report(): Capture {
        if (this.#previousCapture == null) {
            throw new Error("No value captured yet.")
        }
        return this.#previousCapture
    }

    print(format: Format): string {
        return new Formatter(format, this.#config).print(this.report())
    }

    private createCapture(value: any | null | undefined): Capture {
        if (value === null || value === undefined) return new NullCapture();
        else if (typeof value === 'number' && Number.isInteger(value)) return new PrimitiveCapture(CaptureType.INT, false).addValue(this.#config, value);
        else if (typeof value === 'number' && !Number.isInteger(value)) return new PrimitiveCapture(CaptureType.DOUBLE, false).addValue(this.#config, value);
        else if (typeof value === 'boolean') return new PrimitiveCapture(CaptureType.BOOLEAN, false).addValue(this.#config, value);
        else if (typeof value === 'string') return new PrimitiveCapture(CaptureType.TEXT, false).addValue(this.#config, value);
        else if (Array.isArray(value)) {
            if (value.length === 0) {
                return new ListCapture(false, new UnknownCapture())
            } else {
                const subtype = value
                    .map((it) => this.createCapture(it))
                    .reduce((acc, other) => acc.reconcile(this.#config, other));
                return new ListCapture(false, subtype);
            }
        } else if (typeof value === 'object') {
            const fields = new Map<string, Capture>();
            Object.entries(value)
                .forEach(([key, value]) => {
                    fields.set(key, this.createCapture(value))
                });

            return new ObjectCapture(false, fields);
        } else {
            throw new Error(`Unknown node-type: ${typeof value}`)
        }
    }
}