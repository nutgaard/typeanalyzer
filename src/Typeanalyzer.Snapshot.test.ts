import {describe, expect, it} from 'vitest';
import {Typeanalyzer} from "./Typeanalyzer";
import {Formatter, KotlinFormat, TypescriptFormat} from "./Formatter";
import {CaptureType} from "./CaptureType";


interface RootObject {
    id?: string;
    nullValue?: string | null;
    active?: boolean;
    count?: number;
    countLong?: number;
    fraction?: number;
    fractionFloat?: number;
    emptyArray?: Array<Stuff | null | undefined>;
    listOfStuff?: Array<Stuff | null | undefined>;
}

interface Stuff {
    id?: string;
    meta?: { [key: string]: string | undefined | null };
}


const valueObject: RootObject = {
    id: "id",
    nullValue: null,
    active: true,
    count: 123,
    countLong: 1231231,
    fraction: 0.3,
    fractionFloat: 0.333,
    emptyArray: [],
    listOfStuff: [
        {
            id: 'random long string',
            meta: {
                key: "value",
                key2: "other"
            }
        }
    ]
};

describe('Typeanalyzer Snapshot', () => {
    it('should be able to create simple analysis', () => {
        const analyzer = new Typeanalyzer();
        const capture = analyzer.capture(valueObject);

        expect(analyzer.report()).toEqual(capture);
        expect(capture).toMatchSnapshot();
    });

    it('should reconcile nullability of preexisting fields', () => {
        const analyzer = new Typeanalyzer();
        const captures = [
            analyzer.capture(valueObject),
            analyzer.capture({}),
            analyzer.capture(null),
            analyzer.capture(undefined),
            analyzer.capture({ ...valueObject, listOfStuff: [
                    { id: null, meta: { key: null, key2: null }},
                    null,
                    { id: null, meta: null},
            ]})
        ];

        expect(analyzer.report()).not.toEqual(captures[0]);
        expect(analyzer.report()).toEqual(captures[captures.length - 1]);
        for (const capture of captures) {
            expect(capture).toMatchSnapshot();
        }
    });

    it('should update all nullability fields of lists', () => {
        const analyzer = new Typeanalyzer();
        const captures = [
            analyzer.capture({ list: [] }),
            analyzer.capture({ list: ["content"] }),
            analyzer.capture({ list: ["content", null, "more"] }),
            analyzer.capture({ list: null }),
            analyzer.capture(null),
        ];

        expect(analyzer.report()).not.toEqual(captures[0]);
        expect(analyzer.report()).toEqual(captures[captures.length - 1]);
        for (const capture of captures) {
            expect(capture).toMatchSnapshot();
        }
    });

    it('should make key mismatches into nullable elements', () => {
        const analyzer = new Typeanalyzer();
        const captures = [
            analyzer.capture({ key1: 'value' }),
            analyzer.capture({ key2: 'value' }),
        ];

        expect(analyzer.report()).not.toEqual(captures[0]);
        expect(analyzer.report()).toEqual(captures[captures.length - 1]);
        for (const capture of captures) {
            expect(capture).toMatchSnapshot();
        }
    });

    it('should override types even if types mismatch', () => {
        const analyzer = new Typeanalyzer();
        const captures = [
            analyzer.capture(valueObject),
            analyzer.capture({
                active: true,
                count: null,
                countLong: null,
                fraction: null,
            }),
        ];

        expect(analyzer.report()).not.toEqual(captures[0]);
        expect(analyzer.report()).toEqual(captures[captures.length - 1]);
        for (const capture of captures) {
            expect(capture).toMatchSnapshot();
        }
    });

    it('should not throw error', () => {
        const analyzer = new Typeanalyzer();
        analyzer.capture(valueObject);

        const capture = analyzer.capture("a primitive value");

        expect(capture).toMatchSnapshot();
    });

    it('should prettyprint output', () => {
        const analyzer = new Typeanalyzer();
        const capture = analyzer.capture(valueObject);

        expect(analyzer.print(new TypescriptFormat())).toMatchSnapshot();
        expect(new Formatter(new KotlinFormat(), { uniontypeThreshold: 0, captureValuesFor: [], lut: {}}).print(capture)).toMatchSnapshot();
    });

    it('should create uniontype if below threshold', () => {
        const analyzer = new Typeanalyzer({ uniontypeThreshold: 5, captureValuesFor: [CaptureType.TEXT] });
        analyzer.capture({ key: 'value1' });
        analyzer.capture({ key: 'value2' });
        analyzer.capture({ key: 'value3' });
        analyzer.capture({ key: 'value4' });
        analyzer.capture({ key: 'value5' });

        expect(analyzer.print(new TypescriptFormat())).toMatchSnapshot();
    });

    it('should stop capturing data after union threshold is met', () => {
        const analyzer = new Typeanalyzer({ uniontypeThreshold: 2, captureValuesFor: [CaptureType.TEXT] });
        analyzer.capture({ key: 'value1' });
        analyzer.capture({ key: 'value2' });
        analyzer.capture({ key: 'value3' });
        analyzer.capture({ key: 'value4' });
        analyzer.capture({ key: 'value5' });
        analyzer.capture({ key: 'value6' });

        expect(analyzer.report()).toMatchSnapshot();
        expect(analyzer.print(new TypescriptFormat())).toMatchSnapshot()
    });
});