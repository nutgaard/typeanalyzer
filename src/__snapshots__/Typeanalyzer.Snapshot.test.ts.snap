// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Typeanalyzer Snapshot > should be able to create simple analysis 1`] = `
ObjectCapture {
  "fields": Map {
    "id" => PrimitiveCapture {
      "nullable": false,
      "type": "TEXT",
    },
    "nullValue" => NullCapture {
      "type": "NULL",
    },
    "active" => PrimitiveCapture {
      "nullable": false,
      "type": "BOOLEAN",
    },
    "count" => PrimitiveCapture {
      "nullable": false,
      "type": "INT",
    },
    "countLong" => PrimitiveCapture {
      "nullable": false,
      "type": "INT",
    },
    "fraction" => PrimitiveCapture {
      "nullable": false,
      "type": "DOUBLE",
    },
    "fractionFloat" => PrimitiveCapture {
      "nullable": false,
      "type": "DOUBLE",
    },
    "emptyArray" => ListCapture {
      "nullable": false,
      "subtype": UnknownCapture {
        "type": "UNKNOWN",
      },
      "type": "LIST",
    },
    "listOfStuff" => ListCapture {
      "nullable": false,
      "subtype": ObjectCapture {
        "fields": Map {
          "id" => PrimitiveCapture {
            "nullable": false,
            "type": "TEXT",
          },
          "meta" => ObjectCapture {
            "fields": Map {
              "key" => PrimitiveCapture {
                "nullable": false,
                "type": "TEXT",
              },
              "key2" => PrimitiveCapture {
                "nullable": false,
                "type": "TEXT",
              },
            },
            "nullable": false,
            "type": "OBJECT",
          },
        },
        "nullable": false,
        "type": "OBJECT",
      },
      "type": "LIST",
    },
  },
  "nullable": false,
  "type": "OBJECT",
}
`;

exports[`Typeanalyzer Snapshot > should make key mismatches into nullable elements 1`] = `
ObjectCapture {
  "fields": Map {
    "key1" => PrimitiveCapture {
      "nullable": false,
      "type": "TEXT",
    },
  },
  "nullable": false,
  "type": "OBJECT",
}
`;

exports[`Typeanalyzer Snapshot > should make key mismatches into nullable elements 2`] = `
ObjectCapture {
  "fields": Map {
    "key2" => PrimitiveCapture {
      "nullable": true,
      "type": "TEXT",
    },
    "key1" => PrimitiveCapture {
      "nullable": true,
      "type": "TEXT",
    },
  },
  "nullable": false,
  "type": "OBJECT",
}
`;

exports[`Typeanalyzer Snapshot > should not throw error 1`] = `
ObjectCapture {
  "fields": Map {
    "id" => PrimitiveCapture {
      "nullable": false,
      "type": "TEXT",
    },
    "nullValue" => NullCapture {
      "type": "NULL",
    },
    "active" => PrimitiveCapture {
      "nullable": false,
      "type": "BOOLEAN",
    },
    "count" => PrimitiveCapture {
      "nullable": false,
      "type": "INT",
    },
    "countLong" => PrimitiveCapture {
      "nullable": false,
      "type": "INT",
    },
    "fraction" => PrimitiveCapture {
      "nullable": false,
      "type": "DOUBLE",
    },
    "fractionFloat" => PrimitiveCapture {
      "nullable": false,
      "type": "DOUBLE",
    },
    "emptyArray" => ListCapture {
      "nullable": false,
      "subtype": UnknownCapture {
        "type": "UNKNOWN",
      },
      "type": "LIST",
    },
    "listOfStuff" => ListCapture {
      "nullable": false,
      "subtype": ObjectCapture {
        "fields": Map {
          "id" => PrimitiveCapture {
            "nullable": false,
            "type": "TEXT",
          },
          "meta" => ObjectCapture {
            "fields": Map {
              "key" => PrimitiveCapture {
                "nullable": false,
                "type": "TEXT",
              },
              "key2" => PrimitiveCapture {
                "nullable": false,
                "type": "TEXT",
              },
            },
            "nullable": false,
            "type": "OBJECT",
          },
        },
        "nullable": false,
        "type": "OBJECT",
      },
      "type": "LIST",
    },
  },
  "nullable": false,
  "type": "OBJECT",
}
`;

exports[`Typeanalyzer Snapshot > should override types even if types mismatch 1`] = `
ObjectCapture {
  "fields": Map {
    "id" => PrimitiveCapture {
      "nullable": false,
      "type": "TEXT",
    },
    "nullValue" => NullCapture {
      "type": "NULL",
    },
    "active" => PrimitiveCapture {
      "nullable": false,
      "type": "BOOLEAN",
    },
    "count" => PrimitiveCapture {
      "nullable": false,
      "type": "INT",
    },
    "countLong" => PrimitiveCapture {
      "nullable": false,
      "type": "INT",
    },
    "fraction" => PrimitiveCapture {
      "nullable": false,
      "type": "DOUBLE",
    },
    "fractionFloat" => PrimitiveCapture {
      "nullable": false,
      "type": "DOUBLE",
    },
    "emptyArray" => ListCapture {
      "nullable": false,
      "subtype": UnknownCapture {
        "type": "UNKNOWN",
      },
      "type": "LIST",
    },
    "listOfStuff" => ListCapture {
      "nullable": false,
      "subtype": ObjectCapture {
        "fields": Map {
          "id" => PrimitiveCapture {
            "nullable": false,
            "type": "TEXT",
          },
          "meta" => ObjectCapture {
            "fields": Map {
              "key" => PrimitiveCapture {
                "nullable": false,
                "type": "TEXT",
              },
              "key2" => PrimitiveCapture {
                "nullable": false,
                "type": "TEXT",
              },
            },
            "nullable": false,
            "type": "OBJECT",
          },
        },
        "nullable": false,
        "type": "OBJECT",
      },
      "type": "LIST",
    },
  },
  "nullable": false,
  "type": "OBJECT",
}
`;

exports[`Typeanalyzer Snapshot > should override types even if types mismatch 2`] = `
ObjectCapture {
  "fields": Map {
    "active" => PrimitiveCapture {
      "nullable": false,
      "type": "BOOLEAN",
    },
    "count" => PrimitiveCapture {
      "nullable": true,
      "type": "INT",
    },
    "countLong" => PrimitiveCapture {
      "nullable": true,
      "type": "INT",
    },
    "fraction" => PrimitiveCapture {
      "nullable": true,
      "type": "DOUBLE",
    },
    "id" => PrimitiveCapture {
      "nullable": true,
      "type": "TEXT",
    },
    "nullValue" => NullCapture {
      "type": "NULL",
    },
    "fractionFloat" => PrimitiveCapture {
      "nullable": true,
      "type": "DOUBLE",
    },
    "emptyArray" => ListCapture {
      "nullable": true,
      "subtype": UnknownCapture {
        "type": "UNKNOWN",
      },
      "type": "LIST",
    },
    "listOfStuff" => ListCapture {
      "nullable": true,
      "subtype": ObjectCapture {
        "fields": Map {
          "id" => PrimitiveCapture {
            "nullable": false,
            "type": "TEXT",
          },
          "meta" => ObjectCapture {
            "fields": Map {
              "key" => PrimitiveCapture {
                "nullable": false,
                "type": "TEXT",
              },
              "key2" => PrimitiveCapture {
                "nullable": false,
                "type": "TEXT",
              },
            },
            "nullable": false,
            "type": "OBJECT",
          },
        },
        "nullable": false,
        "type": "OBJECT",
      },
      "type": "LIST",
    },
  },
  "nullable": false,
  "type": "OBJECT",
}
`;

exports[`Typeanalyzer Snapshot > should prettyprint output 1`] = `
"interface Generated_691eefc82ffff {
  id: string;
  nullValue: null;
  active: boolean;
  count: number;
  countLong: number;
  fraction: number;
  fractionFloat: number;
  emptyArray: Array<unknown>;
  listOfStuff: Array<Generated_c66d29ffe7fff>;
}

interface Generated_c66d29ffe7fff {
  id: string;
  meta: Generated_16b7135188ffff;
}

interface Generated_16b7135188ffff {
  key: string;
  key2: string;
}
"
`;

exports[`Typeanalyzer Snapshot > should prettyprint output 2`] = `
"data class Generated_691eefc82ffff(
    val id: String,
    val nullValue: Unit?,
    val active: Boolean,
    val count: Int,
    val countLong: Int,
    val fraction: Double,
    val fractionFloat: Double,
    val emptyArray: List<Any>,
    val listOfStuff: List<Generated_c66d29ffe7fff>,
)

data class Generated_c66d29ffe7fff(
    val id: String,
    val meta: Generated_16b7135188ffff,
)

data class Generated_16b7135188ffff(
    val key: String,
    val key2: String,
)
"
`;

exports[`Typeanalyzer Snapshot > should reconcile nullability of preexisting fields 1`] = `
ObjectCapture {
  "fields": Map {
    "id" => PrimitiveCapture {
      "nullable": false,
      "type": "TEXT",
    },
    "nullValue" => NullCapture {
      "type": "NULL",
    },
    "active" => PrimitiveCapture {
      "nullable": false,
      "type": "BOOLEAN",
    },
    "count" => PrimitiveCapture {
      "nullable": false,
      "type": "INT",
    },
    "countLong" => PrimitiveCapture {
      "nullable": false,
      "type": "INT",
    },
    "fraction" => PrimitiveCapture {
      "nullable": false,
      "type": "DOUBLE",
    },
    "fractionFloat" => PrimitiveCapture {
      "nullable": false,
      "type": "DOUBLE",
    },
    "emptyArray" => ListCapture {
      "nullable": false,
      "subtype": UnknownCapture {
        "type": "UNKNOWN",
      },
      "type": "LIST",
    },
    "listOfStuff" => ListCapture {
      "nullable": false,
      "subtype": ObjectCapture {
        "fields": Map {
          "id" => PrimitiveCapture {
            "nullable": false,
            "type": "TEXT",
          },
          "meta" => ObjectCapture {
            "fields": Map {
              "key" => PrimitiveCapture {
                "nullable": false,
                "type": "TEXT",
              },
              "key2" => PrimitiveCapture {
                "nullable": false,
                "type": "TEXT",
              },
            },
            "nullable": false,
            "type": "OBJECT",
          },
        },
        "nullable": false,
        "type": "OBJECT",
      },
      "type": "LIST",
    },
  },
  "nullable": false,
  "type": "OBJECT",
}
`;

exports[`Typeanalyzer Snapshot > should reconcile nullability of preexisting fields 2`] = `
ObjectCapture {
  "fields": Map {
    "id" => PrimitiveCapture {
      "nullable": true,
      "type": "TEXT",
    },
    "nullValue" => NullCapture {
      "type": "NULL",
    },
    "active" => PrimitiveCapture {
      "nullable": true,
      "type": "BOOLEAN",
    },
    "count" => PrimitiveCapture {
      "nullable": true,
      "type": "INT",
    },
    "countLong" => PrimitiveCapture {
      "nullable": true,
      "type": "INT",
    },
    "fraction" => PrimitiveCapture {
      "nullable": true,
      "type": "DOUBLE",
    },
    "fractionFloat" => PrimitiveCapture {
      "nullable": true,
      "type": "DOUBLE",
    },
    "emptyArray" => ListCapture {
      "nullable": true,
      "subtype": UnknownCapture {
        "type": "UNKNOWN",
      },
      "type": "LIST",
    },
    "listOfStuff" => ListCapture {
      "nullable": true,
      "subtype": ObjectCapture {
        "fields": Map {
          "id" => PrimitiveCapture {
            "nullable": false,
            "type": "TEXT",
          },
          "meta" => ObjectCapture {
            "fields": Map {
              "key" => PrimitiveCapture {
                "nullable": false,
                "type": "TEXT",
              },
              "key2" => PrimitiveCapture {
                "nullable": false,
                "type": "TEXT",
              },
            },
            "nullable": false,
            "type": "OBJECT",
          },
        },
        "nullable": false,
        "type": "OBJECT",
      },
      "type": "LIST",
    },
  },
  "nullable": false,
  "type": "OBJECT",
}
`;

exports[`Typeanalyzer Snapshot > should reconcile nullability of preexisting fields 3`] = `
ObjectCapture {
  "fields": Map {
    "id" => PrimitiveCapture {
      "nullable": true,
      "type": "TEXT",
    },
    "nullValue" => NullCapture {
      "type": "NULL",
    },
    "active" => PrimitiveCapture {
      "nullable": true,
      "type": "BOOLEAN",
    },
    "count" => PrimitiveCapture {
      "nullable": true,
      "type": "INT",
    },
    "countLong" => PrimitiveCapture {
      "nullable": true,
      "type": "INT",
    },
    "fraction" => PrimitiveCapture {
      "nullable": true,
      "type": "DOUBLE",
    },
    "fractionFloat" => PrimitiveCapture {
      "nullable": true,
      "type": "DOUBLE",
    },
    "emptyArray" => ListCapture {
      "nullable": true,
      "subtype": UnknownCapture {
        "type": "UNKNOWN",
      },
      "type": "LIST",
    },
    "listOfStuff" => ListCapture {
      "nullable": true,
      "subtype": ObjectCapture {
        "fields": Map {
          "id" => PrimitiveCapture {
            "nullable": false,
            "type": "TEXT",
          },
          "meta" => ObjectCapture {
            "fields": Map {
              "key" => PrimitiveCapture {
                "nullable": false,
                "type": "TEXT",
              },
              "key2" => PrimitiveCapture {
                "nullable": false,
                "type": "TEXT",
              },
            },
            "nullable": false,
            "type": "OBJECT",
          },
        },
        "nullable": false,
        "type": "OBJECT",
      },
      "type": "LIST",
    },
  },
  "nullable": true,
  "type": "OBJECT",
}
`;

exports[`Typeanalyzer Snapshot > should reconcile nullability of preexisting fields 4`] = `
ObjectCapture {
  "fields": Map {
    "id" => PrimitiveCapture {
      "nullable": true,
      "type": "TEXT",
    },
    "nullValue" => NullCapture {
      "type": "NULL",
    },
    "active" => PrimitiveCapture {
      "nullable": true,
      "type": "BOOLEAN",
    },
    "count" => PrimitiveCapture {
      "nullable": true,
      "type": "INT",
    },
    "countLong" => PrimitiveCapture {
      "nullable": true,
      "type": "INT",
    },
    "fraction" => PrimitiveCapture {
      "nullable": true,
      "type": "DOUBLE",
    },
    "fractionFloat" => PrimitiveCapture {
      "nullable": true,
      "type": "DOUBLE",
    },
    "emptyArray" => ListCapture {
      "nullable": true,
      "subtype": UnknownCapture {
        "type": "UNKNOWN",
      },
      "type": "LIST",
    },
    "listOfStuff" => ListCapture {
      "nullable": true,
      "subtype": ObjectCapture {
        "fields": Map {
          "id" => PrimitiveCapture {
            "nullable": false,
            "type": "TEXT",
          },
          "meta" => ObjectCapture {
            "fields": Map {
              "key" => PrimitiveCapture {
                "nullable": false,
                "type": "TEXT",
              },
              "key2" => PrimitiveCapture {
                "nullable": false,
                "type": "TEXT",
              },
            },
            "nullable": false,
            "type": "OBJECT",
          },
        },
        "nullable": false,
        "type": "OBJECT",
      },
      "type": "LIST",
    },
  },
  "nullable": true,
  "type": "OBJECT",
}
`;

exports[`Typeanalyzer Snapshot > should reconcile nullability of preexisting fields 5`] = `
ObjectCapture {
  "fields": Map {
    "id" => PrimitiveCapture {
      "nullable": true,
      "type": "TEXT",
    },
    "nullValue" => NullCapture {
      "type": "NULL",
    },
    "active" => PrimitiveCapture {
      "nullable": true,
      "type": "BOOLEAN",
    },
    "count" => PrimitiveCapture {
      "nullable": true,
      "type": "INT",
    },
    "countLong" => PrimitiveCapture {
      "nullable": true,
      "type": "INT",
    },
    "fraction" => PrimitiveCapture {
      "nullable": true,
      "type": "DOUBLE",
    },
    "fractionFloat" => PrimitiveCapture {
      "nullable": true,
      "type": "DOUBLE",
    },
    "emptyArray" => ListCapture {
      "nullable": true,
      "subtype": UnknownCapture {
        "type": "UNKNOWN",
      },
      "type": "LIST",
    },
    "listOfStuff" => ListCapture {
      "nullable": true,
      "subtype": ObjectCapture {
        "fields": Map {
          "id" => PrimitiveCapture {
            "nullable": true,
            "type": "TEXT",
          },
          "meta" => ObjectCapture {
            "fields": Map {
              "key" => PrimitiveCapture {
                "nullable": true,
                "type": "TEXT",
              },
              "key2" => PrimitiveCapture {
                "nullable": true,
                "type": "TEXT",
              },
            },
            "nullable": true,
            "type": "OBJECT",
          },
        },
        "nullable": true,
        "type": "OBJECT",
      },
      "type": "LIST",
    },
  },
  "nullable": true,
  "type": "OBJECT",
}
`;

exports[`Typeanalyzer Snapshot > should update all nullability fields of lists 1`] = `
ObjectCapture {
  "fields": Map {
    "list" => ListCapture {
      "nullable": false,
      "subtype": UnknownCapture {
        "type": "UNKNOWN",
      },
      "type": "LIST",
    },
  },
  "nullable": false,
  "type": "OBJECT",
}
`;

exports[`Typeanalyzer Snapshot > should update all nullability fields of lists 2`] = `
ObjectCapture {
  "fields": Map {
    "list" => ListCapture {
      "nullable": false,
      "subtype": PrimitiveCapture {
        "nullable": false,
        "type": "TEXT",
      },
      "type": "LIST",
    },
  },
  "nullable": false,
  "type": "OBJECT",
}
`;

exports[`Typeanalyzer Snapshot > should update all nullability fields of lists 3`] = `
ObjectCapture {
  "fields": Map {
    "list" => ListCapture {
      "nullable": false,
      "subtype": PrimitiveCapture {
        "nullable": true,
        "type": "TEXT",
      },
      "type": "LIST",
    },
  },
  "nullable": false,
  "type": "OBJECT",
}
`;

exports[`Typeanalyzer Snapshot > should update all nullability fields of lists 4`] = `
ObjectCapture {
  "fields": Map {
    "list" => ListCapture {
      "nullable": true,
      "subtype": PrimitiveCapture {
        "nullable": true,
        "type": "TEXT",
      },
      "type": "LIST",
    },
  },
  "nullable": false,
  "type": "OBJECT",
}
`;

exports[`Typeanalyzer Snapshot > should update all nullability fields of lists 5`] = `
ObjectCapture {
  "fields": Map {
    "list" => ListCapture {
      "nullable": true,
      "subtype": PrimitiveCapture {
        "nullable": true,
        "type": "TEXT",
      },
      "type": "LIST",
    },
  },
  "nullable": true,
  "type": "OBJECT",
}
`;
